### PR TITLE
Add human-readable __str__ to ValidationError

### PR DESCRIPTION
**Pull request by cjwatson**
When upgraded to zope.schema >= 4.6.0, one of Launchpad's doctests failed with this output, from an API endpoint that printed validation errors on a field named 'tag' when attempting to set it to an invalid value:

    - tag: an invalid tag !!!
    + tag: (u'an invalid tag !!!', 'tag')
    ?      +++                  +++++++++

This is produced by lazr.restful, using code along the lines of:

    try:
        field.validate(value)
    except ValidationError as e:
        errors.append(u"%s: %s" % (name, e))

While I can obviously just change the doctest, neither the old nor the new output is particularly good here.  The old output is neat, but it doesn't actually say what the problem is other than giving the field
name and the objectionable value (is it an unsatisfied constraint, or a wrong type, or some kind of nested validation error, or ...).  The new output is worse: it exposes too much syntax for a good error message, and it duplicates the field name because it's just the stringification of the tuple of all the exception arguments.  Switching to the exception's `repr` would fix the issue that we don't say what the problem is, but in a way that's made up of Python syntax and is intended for developers rather than users.

Although I haven't specifically run into them yet, we may well have similar problems in our web UI.  In that case it will definitely be inappropriate to show the field name (since that would be an identifier
that doesn't necessarily correspond to the field's label in the UI), and again we'll want to avoid excessive Python syntax leaking out.

While we could work around this in lazr.restful and/or Launchpad, it seems better to fix this at its root and give ValidationError instances a proper `__str__` implementation.  The principles I've tried to follow are:

 * Don't render field names, except in some cases of nested validation errors; the frontend is likely to have a better idea of how to indicate which error goes with which field.

 * State the nature of the validation error as well as just its arguments.

 * Minimise code duplication, to make it likely that if somebody adds another ValidationError subclass it will be stringified reasonably by default.

The results aren't perfect: I'd hoped to try to make the `__str__` output be a single line in more cases, but the existing docstrings make that a bit awkward to achieve without parsing translated strings in a
way I don't think would be appropriate.  The rendering of nested errors might also get a bit hard to read in extreme cases.  Nevertheless, I think this is a much better basis for frontends displaying errors to users.

### CODE REVIEW

**Comment by jamadden** [comment]
I certainly appreciate the work that went into this! But it feels to me like it's just about two steps forward, two steps backwards. Perhaps we can brainstorm something that more directly solves the problems you mention.

Even with this change, I don't think these strings are suitable for an end user:

- The strings aren't translated or translatable AFAICS.
- The strings still contain technical jargon like "schema" and implementation details like class and interface names.
- Embedded newlines tend to be lost in certain presentation contexts

And while they get part of the way there for consumers, they do so at the expense of developers:

- As a developer, I *really* want to see the field name if a validation error is logged/printed. (I'd also *really* like to see fully-qualified names of classes and interfaces, but that's another issue.)
- Embedded newlines tend to make grepping or otherwise (even visually) parsing error messages/logs more difficult, especially when there's no structure to them (e.g., indentation of newlines like zope.configuration errors).

Even for developers, I can certainly appreciate how much more information is in these strings! I'd happily give a 👍 if `__str__` kept the field name and didn't introduce extra newlines. Perhaps there's a new method or function that formats an exception on multiple lines?

The expanded exception hierarchy, together with providing additional detail attributes such as `field`, `value`, `schema`, `bound` and so on, was intended to allow applications to be able to produce their own rich, appropriate, presentations for end users. For example, given an application that's rendered entirely in-browser, the server might package the exception as a JSON object (Python dictionary) with corresponding fields (maybe using the exception class name as a "type" field) and let the browser decide exactly how to present the information. Maybe there's something we can do to formalize that at this level? I could see this level being responsible for doing things like translating `TOO_BIG` to `">"`, for example.

**Comment by cjwatson** [comment]
I appreciate that the expanded exception hierarchy is intended to let applications produce their own rich presentations.  The problem is that applications need to know a bit too much about all of it: there's quite a variety of attributes they might need to consider, and no obvious guarantee that they might not need to consider more of them in future.  A gigantic `try`/`except` tree to work out all the details feels like it's at the wrong layer.

Maybe, instead of trying to ram this into `__str__` with all the possible nesting considerations, it would be better to have something like `ValidationError.describe` that returns some kind of structure?  We could still have the property that each error type knows how to render itself to something that's more or less a string rather than a complicated collection of attributes, but in cases where there's some kind of nesting involved, `describe` could return something involving a list or a dict or whatever and the application could work out an appropriate presentation.

I really want to have some way to extract presentational information that doesn't involve the field name; but with this sort of approach maybe it could be a layer above `describe` so that we can both get what we want.

**Comment by cjwatson** [comment]
(And I realise that it would need to be specified more robustly than "a list or a dict or whatever", otherwise we're back to essentially the same problem.)

**Comment by jamadden** [comment]
> Maybe, instead of trying to ram this into __str__ with all the possible nesting considerations, it would be better to have something like `ValidationError.describe` that returns some kind of structure?

I definitely like that idea. I was trying to suggest something similar with the mentions of a method or function or formalizing the dict description. I was incredibly vague though...

> And I realise that it would need to be specified more robustly than "a list or a dict or whatever", otherwise we're back to essentially the same problem

...because I stalled thinking about this problem. I'm not sure what that looks like besides just dumping `str` of each of the various attributes into a dict (or, if the attribute happens to be a list, like `errors` for SchemaValidationError`, or another exception instance, recursing). That's pretty much what our code that renders these to JSON does, but I didn't know if that's helpful generally. If it is, maybe the protocol here is just to implement `__json__`? (One thing our code handles specially: Some validation errors like those from z3c.password define an `i18n_message` attribute that the server takes care of translating.)

> I appreciate that the expanded exception hierarchy is intended to let applications produce their own rich presentations. The problem is that applications need to know a bit too much about all of it

Absolutely, I can see how in general that could be a problem. I guess in our applications we haven't really run into that because the context helps guide the UI about what to expect, and when there's no helpful context it's worked OK to only do something special for the most common types of errors and otherwise be a bit more generic.

File: src/zope/schema/_bootstrapinterfaces.py [hunk 1]
```diff
@@ -165,12 +204,18 @@ class TooLong(LenOutOfBounds):
 
     violation_direction = OutOfBounds.TOO_LARGE
 
+    def _describe(self):
+        return 'len(%r) > %s' % (self.value, self.bound)
```
Comment (mgedmin) [inline src/zope/schema/_bootstrapinterfaces.py]: I thought you didn't want excessive Python syntax leaking out? ;)

File: src/zope/schema/_bootstrapinterfaces.py [hunk 2]
```diff
@@ -230,10 +278,23 @@ def __init__(self, errors=None, name=None, schema_errors=None, invariant_errors=
...
 
+    def _describe(self):
+        error_strs = []
+        if self.schema_errors:
+            error_strs.extend([
+                '%s: %s' % (name, error)
+                for name, error in sorted(self.schema_errors.items())])
+        if self.invariant_errors:
+            error_strs.extend([str(error) for error in self.invariant_errors])
+        return '\n'.join(error_strs)
```
Comment (mgedmin) [inline src/zope/schema/_bootstrapinterfaces.py]: I worry that embedded newlines inside each field's error will make this harder to read.

I would be tempted to fix that by doing `'%s: %s' % (name, str(error).replace('\n', '\n  '))` to preserve the hierarchy with indentation.  That should suffice to make it clearer which error refers to which field in text logs even if you have multiple levels of nesting.

**Comment by mgedmin** [comment]
> I appreciate that the expanded exception hierarchy is intended to let applications produce their own rich presentations. The problem is that applications need to know a bit too much about all of it: there's quite a variety of attributes they might need to consider, and no obvious guarantee that they might not need to consider more of them in future. A gigantic try/except tree to work out all the details feels like it's at the wrong layer.

From what I know about the creators of zope.schema, the intended approach was to register custom components (e.g. snippet views, but more likely viewlets) for each error type you want to present in your UI.  Gigantic try/except trees are a strawman.

The rest is spot on: an app needs to know all the error types it wants to present nicely.  It would be nice if that weren't so.  I'm not sure that's solvable.

**Comment by icemac** [comment]
What is open before this PR can be merged?

**Comment by jamadden** [comment]
There are open questions/fixes with newlines and field names.

**Comment by cjwatson** [comment]
I'm going to withdraw this, at least for the moment.  I'm afraid I don't have the energy to work out how to make it acceptable.  I don't mind if somebody else wants to pick it up (or not).

**Pull request closed without merging**